dex/testing/eth/harness.sh

```diff
@@ -0,0 +1,139 @@
+ALPHA_NODE_PORT="30302"
+
+BETA_ADDRESS="4f8ef3892b65ed7fc356ff473a2ef2ae5ec27a06"
+BETA_ADDRESS_JSON_FILE_NAME="UTC--2021-01-27T08-20-58.179642501Z--4f8ef3892b65ed7fc356ff473a2ef2ae5ec27a06"
+BETA_ADDRESS_JSON='{"address":"4f8ef3892b65ed7fc356ff473a2ef2ae5ec27a06","crypto":{"cipher":"aes-128-ctr","ciphertext":"c5672bb829df9e209ca8ce18dbdd1fed69c603d639e06ab09127b672a609c121","cipherparams":{"iv":"24460eb2934c8b61cee3ad0aa7b843c0"},"kdf":"scrypt","kdfparams":{"dklen":32,"n":262144,"p":1,"r":8,"salt":"1f85da881994ca7b4a23f0698da70500a4b79f97a4450b83b129ebf3b4c28f50"},"mac":"1ecea707f1bffa1f6f944cb47e83118d8179e8a5005b83c88610b7e8692a1197"},"id":"56633762-6fb1-4cbf-8396-3a2e4661f7d4","version":3}'
+BETA_NODE_KEY="0f3f23a0f14202da009bd59a96457098acea901986629e54d5be1eea32fc404a"
+BETA_ENODE="b1d3e358ee5c9b268e911f2cab47bc12d0e65c80a6d2b453fece34facc9ac3caed14aa3bc7578166bb08c5bc9719e5a2267ae14e0b42da393f4d86f6d5829061"
+BETA_NODE_PORT="30303"
+
+# PASSWORD is the password used to unlock all accounts/wallets/addresses.
+PASSWORD="abc"
+
+export NODES_ROOT=~/dextest/eth
+export GENESIS_JSON_FILE_LOCATION="${NODES_ROOT}/genesis.json"
+export MINE_JS="${NODES_ROOT}/mine.js"
+
+if [ -d "${NODES_ROOT}" ]; then
+  rm -R "${NODES_ROOT}"
+fi
+
+mkdir -p "${NODES_ROOT}/alpha"
+mkdir -p "${NODES_ROOT}/beta"
+mkdir -p "${NODES_ROOT}/harness-ctl"
+
+echo "Writing ctl scripts"
+################################################################################
+# Control Scripts
+################################################################################
+
+# Write genesis json. ".*Block" fields represent block height where certain
+# protocols take effect. "clique" is our proof of authority scheme. One block
+# can be mined per second with the a signature belonging to the address in
+# "extradata". The addresses in the "alloc" field are allocated "balance".
+cat > "${NODES_ROOT}/genesis.json" <<EOF
+{
+  "config": {
+    "chainId": 42,
+    "homesteadBlock": 0,
+    "eip150Block": 0,
+    "eip155Block": 0,
+    "eip158Block": 0,
+    "byzantiumBlock": 0,
+    "constantinopleBlock": 0,
+    "petersburgBlock": 0,
+    "clique": {
+      "period": 1,
+      "epoch": 30000
+    }
+  },
+  "difficulty": "1",
+  "gasLimit": "8000000",
+  "extradata": "0x00000000000000000000000000000000000000000000000000000000000000009ebba10a6136607688ca4f27fab70e23938cd0270000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
+  "alloc": {
+    "18d65fb8d60c1199bb1ad381be47aa692b482605": {
+        "balance": "1000000000000"
+    },
+    "4f8ef3892b65ed7fc356ff473a2ef2ae5ec27a06": {
+        "balance": "1000000000000"
+    }
+  }
+}
+EOF
+
+# Write mining javascript.
+# NOTE: This sometimes mines more than one block. It is a race.
+cat > "${MINE_JS}" <<EOF
+function mine() {
+  miner.start();
+  miner.stop();
+  admin.sleep(1.1);
+  return true;
+}
+EOF
```

Comment:
This could definitely improved, but I currently can't see how. I don't think there's a way to specify the number of blocks you want to mine. calling start and stop in succession usually mines one, but sometimes, rarley, it mines two. Also, mining will only happen every second. So, if you were to call start stop start stop without waiting, it would only mine one block (or maybe two...) Sleeping for 1.1 secounds seems to ensure that you mine at least the number of blocks you wanted, but could be more.

Reply:
0 also possible
```
[harness-ctl]$ ./mine-alpha 6
1
1
1
0
2
1
```

Reply:
Dang.  There's gotta be a js while loop solution to spin until a block is mined.